pkg/yurthub/healthchecker/health_checker.go

```diff
@@ -21,114 +21,368 @@ import (
+				return
+			}
+			hcm.checkers[server.String()] = checker
 		}
-		hcm.checkers[server.String()] = checker
 	}
 
-	return hcm, nil
+	if err := hcm.storageLease(hcm.latestLease); err != nil {
+		klog.Errorf("storage lease fail: %v", err)
+	}
+
+	go hcm.healthzCheckLoop(hcm.stopCh)
+
+	return
 }
 
-// IsHealthy returns the healthy stats of specified server
-func (hcm *healthCheckerManager) IsHealthy(server *url.URL) bool {
-	hcm.RLock()
-	defer hcm.RUnlock()
-	if checker, ok := hcm.checkers[server.String()]; ok {
-		return checker.isHealthy()
+func (hcm *healthCheckerManager) healthzCheckLoop(stopCh <-chan struct{}) {
+	intervalTicker := time.NewTicker(heartbeatFrequency)
+	defer intervalTicker.Stop()
+
+	for {
+		select {
+		case <-stopCh:
+			klog.Infof("exit normally in health check loop.")
+			return
+		case <-intervalTicker.C:
+			hcm.sync()
+		}
 	}
+}
 
-	return false
+func (hcm *healthCheckerManager) sync() {
+	defer hcm.storageLease(hcm.latestLease)
+	for i := 0; i < len(hcm.remoteServers); i++ {
+		checker := hcm.getChecker()
+		if !hcm.syncWithChecker(checker) {
+			checker.healthyCnt = 0
+			if checker.clusterHealthy {
+				now := time.Now()
+				klog.Infof("cluster becomes unhealthy from %v, healthy status lasts %v, remote server: %v", time.Now(), now.Sub(checker.lastTime), checker.netAddress)
+				checker.setHealthy(false, now)
+			}
+		} else {
+			checker.healthyCnt++
+			if !checker.clusterHealthy && checker.healthyCnt >= checker.healthyThreshold {
+				now := time.Now()
+				klog.Infof("cluster becomes healthy from %v, unhealthy status lasts %v, remote server: %v", now, now.Sub(checker.lastTime), checker.netAddress)
+				checker.setHealthy(true, now)
+			}
+			break
+		}
+	}
 }
 
-type checker struct {
-	sync.RWMutex
-	serverHealthzAddr string
-	healthzClient     *http.Client
-	clusterHealthy    bool
-	lastTime          time.Time
-	onFailureFunc     func(string)
-	netAddress        string
-	failedRetry       int
-	healthyThreshold  int
+func (hcm *healthCheckerManager) syncWithChecker(checker *checker) bool {
+	if hcm.latestLease != nil {
+		err := hcm.retryUpdateLease(hcm.newLease(hcm.latestLease, checker), checker)
+		if err == nil {
+			return true
+		}
+		klog.Infof("failed to update lease using latest lease, fallback to ensure lease, err: %v", err)
+		return false
+	}
+
+	lease, created := hcm.backoffEnsureLease(checker)
+	if lease == nil {
+		return false
+	}
+
+	hcm.setLatestLease(lease)
+
+	if !created {
+		if err := hcm.retryUpdateLease(lease, checker); err != nil {
+			klog.Errorf("%v", err)
+			return false
+		}
+	}
+	return true
 }
 
-func newChecker(url *url.URL, tp transport.Interface, failedRetry, healthyThreshold int, stopCh <-chan struct{}) (*checker, error) {
-	serverHealthzURL := *url
-	if serverHealthzURL.Path == "" || serverHealthzURL.Path == "/" {
-		serverHealthzURL.Path = "/healthz"
+func (hcm *healthCheckerManager) backoffEnsureLease(checker *checker) (*coordinationv1.Lease, bool) {
```

Comment:
Please define Lease process in a interface

Reply:
fixed